virtio-spec/src/virtq/mod.rs

```diff
@@ -0,0 +1,255 @@
+        let ptr = NonNull::new(ptr).unwrap().cast::<le16>();
+        NonNull::slice_from_raw_parts(ptr, len)
+    }
+
+    pub fn ring(&self, has_event_idx: bool) -> &[le16] {
+        let ptr = Self::ring_ptr(NonNull::from(self), has_event_idx);
+        unsafe { ptr.as_ref() }
+    }
+
+    pub fn ring_mut(&mut self, has_event_idx: bool) -> &mut [le16] {
+        let mut ptr = Self::ring_ptr(NonNull::from(self), has_event_idx);
+        unsafe { ptr.as_mut() }
+    }
+
+    pub fn used_event_ptr(this: NonNull<Self>, has_event_idx: bool) -> Option<NonNull<le16>> {
+        if !has_event_idx {
+            return None;
+        }
+
+        let ptr = unsafe { addr_of_mut!((*this.as_ptr()).ring_and_used_event) };
+        let len = ptr.len();
+
+        if len == 0 {
+            return None;
+        }
+
+        let ptr = NonNull::new(ptr).unwrap().cast::<le16>();
+        let ptr = unsafe { ptr.add(len - 1) };
+        Some(ptr)
+    }
+
+    pub fn used_event(&self, has_event_idx: bool) -> Option<&le16> {
+        Self::used_event_ptr(NonNull::from(self), has_event_idx).map(|ptr| unsafe { ptr.as_ref() })
+    }
+
+    pub fn used_event_mut(&mut self, has_event_idx: bool) -> Option<&mut le16> {
+        Self::used_event_ptr(NonNull::from(self), has_event_idx)
+            .map(|mut ptr| unsafe { ptr.as_mut() })
+    }
+}
+
+endian_bitflags! {
+    /// Virtqueue available ring flags
+    #[doc(alias = "VIRTQ_AVAIL_F")]
+    pub struct AvailF: le16 {
+        /// The driver uses this in avail->flags to advise the device: don’t
+        /// interrupt me when you consume a buffer.  It’s unreliable, so it’s
+        /// simply an optimization.
+        #[doc(alias = "VIRTQ_AVAIL_F_NO_INTERRUPT")]
+        const NO_INTERRUPT = 1;
+    }
+}
+
+/// The Virtqueue Used Ring
+#[doc(alias = "virtq_used")]
+#[derive(Debug)]
+#[repr(C)]
+#[repr(align(4))] // mem::align_of::<UsedElem>
+pub struct Used {
+    pub flags: UsedF,
+    pub idx: le16,
+    ring_and_avail_event: [le16],
+}
+
+impl Used {
+    pub fn layout(queue_size: u16, has_event_idx: bool) -> Layout {
+        let mut layout = Layout::array::<le16>(2)
+            .unwrap()
+            .extend(Layout::array::<UsedElem>(queue_size.into()).unwrap())
+            .unwrap()
+            .0;
+
+        if has_event_idx {
+            layout = layout
+                .extend(Layout::new::<le16>())
+                .unwrap()
+                .0
+                .pad_to_align();
+        }
+
+        layout
+    }
+
+    pub fn from_ptr(ptr: NonNull<[u8]>, has_event_idx: bool) -> Option<NonNull<Self>> {
+        let len = ptr.len();
+        let ptr = ptr.cast::<u8>().as_ptr();
+
+        if !ptr.cast::<UsedElem>().is_aligned() {
+            return None;
+        }
+
+        if len % mem::size_of::<le16>() != 0 {
+            return None;
+        }
+
+        let len = len / mem::size_of::<le16>() - 2 - usize::from(has_event_idx);
```

Comment:
Should this subtraction be done? If `avail_event` exists, it is part of the slice.

Reply:
I think we should include a check similar to the one above to see if this `len` is a multiple of (`size_of::<UsedElem> / size_of::<le16>`).

Reply:
> Should this subtraction be done? If `avail_event` exists, it is part of the slice.

Yes, this is subtracting the padding.

> I think we should include a check similar to the one above to see if this `len` is a multiple of (`size_of::<UsedElem> / size_of::<le16>`).

I don't understand. The slice contains `avail_event` after all.

Reply:
I meant something like
```rust
if (len - usize::from(has_event_idx)) % (mem::size_of::<UsedElem>() / mem::size_of::<le16>()) != 0 {
    return None;
}
```
to make sure the part of the slice that is supposed to be `[UsedElem]` actually has a size that is a multiple of `mem::size_of::<UsedElem>()`.

Reply:
I have replaced the previous check with

```rust
if len % mem::size_of::<UsedElem>() != usize::from(!has_event_idx) * mem::size_of::<le32>() {
    return None;
}
```